lib/connections/limiter.go

```diff
@@ -124,6 +127,91 @@ func (lim *limiter) VerifyConfiguration(from, to config.Configuration) error {
 	return nil
 }
 
+func timeDiff(now time.Time, day, hour, minute int) time.Duration {
+	then := time.Date(now.Year(), now.Month(), day, hour, minute, 0, 0, now.Location())
+	return then.Sub(now)
+}
+
+func correspondingEndDay(beginDay, beginHour, beginMinute, hour, minute int) int {
+	if hour < beginHour || (hour == beginHour && minute < beginMinute) {
+		return beginDay + 1
+	}
+
+	return beginDay
+}
+
+func nextChange(beginHour, beginMinute, endHour, endMinute int) time.Duration {
+	beginDay := time.Now().Day()
+	tillBegin := timeDiff(time.Now(), beginDay, beginHour, beginMinute)
+
+	if tillBegin >= 0 {
+		return tillBegin
+	}
+
+	endDay := correspondingEndDay(beginDay, beginHour, beginMinute, endHour, endMinute)
+
+	tillEnd := timeDiff(time.Now(), endDay, endHour, endMinute)
+	if tillEnd >= 0 {
+		return tillEnd
+	}
+
+	beginDay += 1
+
+	return timeDiff(time.Now(), beginDay, beginHour, beginMinute)
+
+}
+
+func (lim *limiter) changeLimits(sendLimitValue, recvLimitValue int) {
+	sendLimitStr := "is unlimited"
+	recvLimitStr := "is unlimited"
+
+	// The rate variables are in KiB/s in the config (despite the camel casing
+	// of the name). We multiply by 1024 to get bytes/s.
+	if recvLimitValue <= 0 {
+		lim.read.SetLimit(rate.Inf)
+	} else {
+		lim.read.SetLimit(1024 * rate.Limit(recvLimitValue))
+		recvLimitStr = fmt.Sprintf("limit is %d KiB/s", recvLimitValue)
+	}
+
+	if sendLimitValue <= 0 {
+		lim.write.SetLimit(rate.Inf)
+	} else {
+		lim.write.SetLimit(1024 * rate.Limit(sendLimitValue))
+		sendLimitStr = fmt.Sprintf("limit is %d KiB/s", sendLimitValue)
+	}
+
+	l.Infof("Overall send rate %s, receive rate %s", sendLimitStr, recvLimitStr)
+}
+
+func (lim *limiter) timerLoop(timer *time.Timer, to config.Configuration) {
+	for range timer.C {
+		lim.mu.Lock()
+		defer lim.mu.Unlock()
+
+		now := time.Now()
+
+		beginDay := now.Day()
+		beginHour := to.Options.ScheduledRatesFromHour
+		beginMinute := to.Options.ScheduledRatesFromMinute
+		begin := time.Date(now.Year(), now.Month(), beginDay, beginHour, beginMinute, 0, 0, now.Location())
+
+		endHour := to.Options.ScheduledRatesToHour
+		endMinute := to.Options.ScheduledRatesToMinute
+		endDay := correspondingEndDay(beginDay, beginHour, beginMinute, endHour, endMinute)
+		end := time.Date(now.Year(), now.Month(), endDay, endHour, endMinute, 0, 0, now.Location())
+
+		if now.Sub(begin) >= 0 && end.Sub(now) >= 0 {
```

Comment:
This time handling duplicates that in `nextChange`: In `nextChange` you already implicitly determine whether the change is from scheduled to normal or the other way around. You could return that as a bool and use it to decide to which limit to change to. Thus no time handling necessary at all here in `timerLoop`.

Reply:
Changed as suggested